### ISSUE TITLE
Feature Request: support sending IXFR transactions after notifying masters

### ISSUE THREAD

**Issue by m-barthelemy** [PowerDNS/pdns#6679]
- Program: Authoritative
 - Issue type: Feature request

### Short description
After receiving a notification, a slave may want to ask a PowerDNS master for zone changes using IXFR.
As of 4.2.3, PowerDNS can only emit AXFR responses, according to the doc and confirmed after a few quick tests with an Unbound slave.

### Usecase
Maintaining slaves up to date almost in real time using NOTIFY is great. However:
 - If the zone becomes big, any update potentially requires transferring (and then updating) a lot of information on the salves.
 - If the slave expects an IXFR response from the master (Unbound 1.7.1 with `auth-zone`, `allow-notify` and `master` configured) if fails. Well, I guess Unbound should be able to accept an AXFR response, and its current behavior is a limitation or a bug.

**Comment by Habbie** [comment]
Hello, I took the liberty of editing the title. This is a big feature request, to be clear, and I'm surprised we did not already have a ticket for it!

**Comment by Habbie** [comment]
You might be interested in [ixfrdist](https://doc.powerdns.com/authoritative/manpages/ixfrdist.1.html), by the way.

**Comment by Habbie** [comment]
> Well, I guess Unbound should be able to accept an AXFR response, and its current behavior is a limitation or a bug.

That does sound like a bug, to be clear.

**Comment by pieterlexis** [comment]
Hi @m-barthelemy,

Due to the use of the database in the backend (and the design thereof) it is quite hard to create an IXFR. This is mostly because we don't keep multiple versions of zones and it is hard to determine removals without many extra tables to keep this history and all the code that comes with it.

In master, there is a tool called [`ixfrdist`](https://doc.powerdns.com/authoritative/manpages/ixfrdist.1.html) ()[config file format](https://doc.powerdns.com/authoritative/manpages/ixfrdist.yml.5.html), currently available in the `pdns-tools` package from [our repos](https://repo.powerdns.com). This tools retrieves the zones from the auth, saves them to disk and re-distributes it over AXFR and IXFR. It is somewhat of a work in progress is only checks for new data at the master every SOA Expire seconds and it does not support NOTIFY or TSIG.

If you want, we can consider this issue a request for NOTIFY support in `ixfrdist`.

> If the slave expects an IXFR response from the master (Unbound 1.7.1 with auth-zone, allow-notify and master configured) if fails. Well, I guess Unbound should be able to accept an AXFR response, and its current behavior is a limitation or a bug.

That _smells_ like a bug, an IXFR request might be answered with a full zone according to [RFC 1995 section 4](https://tools.ietf.org/html/rfc1995#section-4):

>    If incremental zone transfer is not available, the entire zone is returned.  The first and the last RR of the response is the SOA record of the zone.  I.e. the behavior is the same as an AXFR response except the query type is IXFR.

Cheers!

**Comment by Habbie** [comment]
> That does sound like a bug, to be clear.

I cannot reproduce your Unbound problem.

**Comment by qrtp** [comment]
Is it true that PowerDNS does not support IXFR out of the box? I notice that any time a secondary initiates a zone IXFR with a serial from the past, the request always falls back to AXFR and sends the whole zone. This is not idea for large zones.

**Comment by phonedph1** [comment]
> Is it true that PowerDNS does not support IXFR out of the box?

Correct, you could use `ixfrdist` (mentioned above), https://doc.powerdns.com/lightningstream/latest/index.html, or some other database replication if full AXFR does not scale to your size (although it scales to numbers much higher than you may think).

**Comment by qrtp** [comment]
> Correct, you could use `ixfrdist` (mentioned above), https://doc.powerdns.com/lightningstream/latest/index.html, or some other database replication if full AXFR does not scale to your size (although it scales to numbers much higher than you may think).

Thanks for the speedy confirmation. Our zones will have millions of records, and our POC testing shows the server struggles with replication at this scale. I'll definitely look into some of the options you provided.

**Comment by franklouwers** [comment]
Especially at that amount of records, I would look at database replication. which can be "traditional db replication" but also something like Lightning Stream which replicates LMDB databases via S3(-compatible) storage. https://doc.powerdns.com/lightningstream/latest/index.html

**Comment by qrtp** [comment]
@phonedph1 @franklouwers I have `ixfrdist` working very nicely now to provide the incremental transfers. Does this process support TSIG to validate the zone transfers? I couldn't find anything in the YAML documentation, only the ACL of approved IP/network addresses.

For our use case, we unfortunately cannot use any of the optimized DB replication. Our secondary DNS needs to rely upon the standard AXFR/IXFR transfer protocols to ship the data.

**Comment by klaus-nicat** [comment]
> You might be interested in [ixfrdist](https://doc.powerdns.com/authoritative/manpages/ixfrdist.1.html), by the way.

I wonder is there any advantage of using ixfrdist instead of just using an IXFR capable nameserver (Bind, Knot...) between the PowerDNS Primary and the Secondary?
